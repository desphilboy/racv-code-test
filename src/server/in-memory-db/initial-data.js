export const POIN_COOK_HOUSES = [
	{
		state: "VIC",
		zip: "3030",
		suburb: "Point Cook",
		address: "31 Terrene Terrace",
		salePrice: "700000",
		description: "description of the house",
	},
];

export const TARNEIT_HOUSES = [
	{
		state: "VIC",
		zip: "3029",
		suburb: "Tarneit",
		address: "89 Penrose Promenade",
		salePrice: "650000",
		description: "description of the house",
	},
];

export default POIN_COOK_HOUSES;
