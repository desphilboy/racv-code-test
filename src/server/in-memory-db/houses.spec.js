import Houses, { PARTS } from "./houses";
import { POIN_COOK_HOUSES, TARNEIT_HOUSES } from "./initial-data";

describe("In Memory DB", () => {
	describe("Houses Table", () => {
		it("is created without problem", () => {
			expect(Houses).toMatchSnapshot();
		});

		it("can be initialized by empty data", () => {
			Houses.init();
			expect(Houses.findAll()).toEqual([]);
		});

		it("data can be inserted in the table", () => {
			Houses.init();

			Houses.insert(WERRIBEE_HOUSE);

			expect(Houses.findAll()).toEqual([
				{ ...WERRIBEE_HOUSE, id: expect.any(String) },
			]);
		});

		it("insert with empty suburb throws error and is not accomplished", () => {
			Houses.init();

			let error = null;
			try {
				Houses.insert(INVALID_HOUSE);
			} catch (e) {
				error = e;
			}

			expect(error).toEqual({
				code: 501,
				message: "cannot insert house without suburb",
			});

			expect(Houses.findAll()).toEqual([]);
		});

		it("insert with existing suburb adds to that suburb", () => {
			Houses.init(POIN_COOK_HOUSES);

			Houses.insert(NEW_POINTCOOK_HOUSE);

			expect(PARTS["Point Cook"]).toEqual([
				{ ...POIN_COOK_HOUSES[0], id: expect.any(String) },
				{ ...NEW_POINTCOOK_HOUSE, id: expect.any(String) },
			]);
		});

		it("is initialized with initial data", () => {
			Houses.init([...POIN_COOK_HOUSES, ...TARNEIT_HOUSES]);
			expect(Houses.findAll()).toEqual([
				{ ...POIN_COOK_HOUSES[0], id: expect.any(String) },
				{ ...TARNEIT_HOUSES[0], id: expect.any(String) },
			]);
		});

		it("FindAllBySuburb returns houses in a suburb", () => {
			Houses.init([...POIN_COOK_HOUSES, ...TARNEIT_HOUSES]);
			expect(Houses.findAllBySuburb("Point Cook")).toEqual([
				{ ...POIN_COOK_HOUSES[0], id: expect.any(String) },
			]);
			expect(Houses.findAllBySuburb("Mitchem")).toEqual([]);
		});

		it("FindAllBySuburb returns all if suburb is falsy", () => {
			Houses.init([...POIN_COOK_HOUSES, ...TARNEIT_HOUSES]);

			expect(Houses.findAllBySuburb("")).toEqual(Houses.findAll());
		});
	});
});

const WERRIBEE_HOUSE = {
	suburb: "Werribee",
	zip: "3030",
	state: "VIC",
	address: "56 Synot street",
	salePrice: "1800000",
};

const INVALID_HOUSE = {
	zip: "3030",
	state: "VIC",
	address: "56 Synot street",
	salePrice: "1800000",
};

const NEW_POINTCOOK_HOUSE = {
	zip: "3030",
	state: "VIC",
	address: "1 Donaldson Place",
	salePrice: "760000",
	suburb: "Point Cook",
};
