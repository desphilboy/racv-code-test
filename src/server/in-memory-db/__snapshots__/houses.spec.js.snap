// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`In Memory DB Houses Table is created without problem 1`] = `
Object {
  "findAll": [Function],
  "findAllBySuburb": [Function],
  "init": [Function],
  "insert": [Function],
}
`;
